### PR TITLE
nix-hash: support base-64 and SRI format

### PR DESCRIPTION
**Pull request by ShamrockLee**
# Motivation

This PR provides `nix-hash` with base-64 and SRI format support, equivalent to that of `nix hash`.

If applied, users with a `nix-channel`-based setup would be able to compute hashes in base-64 and SRI format without having to type `nix --extra-experimental-features hash ...`

```console
$ nix-hash --type sha256 --base64 /nix/store/g2m8kfw7kpgpph05v2fxcx4d5an09hl3-hello-2.12.1
p7JaKoco7c2MM+1zHW1B5ebUUnt9rGVc3GaTsu1QYfE=

$ nix-hash --type sha256 --sri /nix/store/g2m8kfw7kpgpph05v2fxcx4d5an09hl3-hello-2.12.1
sha256-p7JaKoco7c2MM+1zHW1B5ebUUnt9rGVc3GaTsu1QYfE=

$ nix-hash --type sha1 --to-base64 e4fd8ba5f7bbeaea5ace89fe10255536cd60dab6
5P2Lpfe76upazon+ECVVNs1g2rY=

$ nix-hash --type sha1 --to-sri nvd61k9nalji1zl9rrdfmsmvyyjqpzg4
sha1-5P2Lpfe76upazon+ECVVNs1g2rY=

$ nix-hash --to-base16 sha1-5P2Lpfe76upazon+ECVVNs1g2rY=
e4fd8ba5f7bbeaea5ace89fe10255536cd60dab6
```

# Context

Solves #7688

# Checklist for maintainers

Maintainers: tick if completed or explain if not relevant

 - [ ] agreed on idea
 - [ ] agreed on implementation strategy
 - [ ] tests, as appropriate
   - functional tests - `tests/**.sh`
   - unit tests - `src/*/tests`
   - integration tests
 - [ ] documentation in the manual
 - [ ] code and comments are self-explanatory
 - [ ] commit message explains why the change was made
 - [ ] new feature or bug fix: updated release notes

### CODE REVIEW

**Comment by edolstra** [comment]
I don't think we should be extending `nix-hash` and the old CLI in general. `nix hash` works fine.

**Comment by vcunat** [comment]
I think we're being a bit schizophrenic in (not) recommending the new `nix` CLI.  It's hard to use by default (`--extra-experimental-features nix-command`) and we still say that it can change incompatibly.

For example what would you use in a script?  A script-local alias that adds the extra enabler and additionally hope that you won't be hit by future incompatibility?

**Comment by Ericson2314** [comment]
Discussed in the Nix team meeting:

- @edolstra: Non-zero cost, spent on legacy CLI
- @ericson2314: Nice to merge this and new CLI being stable will be a nice surprise
- @roberth: Needs tests and release notes
- @thufschmitt: Let's ask author to do those. If they do those
- @thufschmit: Assigned

**Comment by ShamrockLee** [comment]
> * @roberth: Needs tests and release notes
> 
> * @thufschmitt: Let's ask author to do those. If they do those

Done.

**Comment by ShamrockLee** [comment]
Oops! I didn't read the hacking guide carefully enough.

Now the test should pass.

File: doc/manual/src/release-notes/rl-next.md [hunk 1]
```diff
@@ -4,6 +4,19 @@
   passed the `--stdin` flag. This is primarily useful when you have a large amount of paths which
   exceed the OS arg limit.
 
+* The `builtins.readDir` function has been optimized when encountering not-yet-known
+  file types from POSIX's `readdir`. In such cases the type of each file is/was
+  discovered by making multiple syscalls. This change makes these operations
+  lazy such that these lookups will only be performed if the attribute is used.
+  This optimization affects a minority of filesystems and operating systems.
+
```
Comment (Ericson2314) [inline doc/manual/src/release-notes/rl-next.md:12]: I think this was from a previous release and you accidentally kept it as part of the merge conflict.

File: tests/hash.sh [hunk 2]
```diff
@@ -28,16 +34,26 @@ try sha256 "abcdbcdecdefdefgefghfghighijhijkijkljklmklmnlmnomnopnopq" "248d6a61d
...
 try sha256 "abcdbcdecdefdefgefghfghighijhijkijkljklmklmnlmnomnopnopq" "sha256-JI1qYdIGOLjlwCaTDD5gOaM85Flk/yFn9uzt1BnbBsE="
+unset FORMAT_FLAG
+
+NO_TEST_NIX_COMMAND=1 # nix-hash [--flat] defaults to the Base16 format
+try sha512 "abc" "ddaf35a193617abacc417349ae20413112e6fa4e89a97ea20a9eeee64b55d39a2192992a274fc1a836ba3c23a3feebbd454d4423643ce80e2a9ac94fa54ca49f"
+unset NO_TEST_NIX_COMMAND
+
+NO_TEST_CLASSIC=1 # nix hash [file|path] defaults to the SRI format
+try sha512 "abc" "sha512-3a81oZNherrMQXNJriBBMRLm+k6JqX6iCp7u5ktV05ohkpkqJ0/BqDa6PCOj/uu9RU1EI2Q86A4qmslPpUyknw=="
+unset NO_TEST_CLASSIC
```
Comment (Ericson2314) [inline tests/hash.sh:56]: We can just set these variables in the single command where it matters, perhaps?
```suggestion
# nix-hash [--flat] defaults to the Base16 format
NO_TEST_NIX_COMMAND=1 try sha512 "abc" "ddaf35a193617abacc417349ae20413112e6fa4e89a97ea20a9eeee64b55d39a2192992a274fc1a836ba3c23a3feebbd454d4423643ce80e2a9ac94fa54ca49f"

# nix hash [file|path] defaults to the SRI format
NO_TEST_CLASSIC=1 try sha512 "abc" "sha512-3a81oZNherrMQXNJriBBMRLm+k6JqX6iCp7u5ktV05ohkpkqJ0/BqDa6PCOj/uu9RU1EI2Q86A4qmslPpUyknw=="
```

File: doc/manual/src/release-notes/rl-next.md (hunk 1 above)
Comment (ShamrockLee) [reply doc/manual/src/release-notes/rl-next.md:12]: There was a merge conflict which I solved by "keeping 'em both".

Comment (Ericson2314) [reply doc/manual/src/release-notes/rl-next.md:12]: Yeah just keep the new one. The merge conflicts in this file are annoying and misleading, sad to say. We should adopt a different system in order to make this easier.

**Pull request merged**